File: C#/AnaliseDeSentimento/AnaliseDeSentimento/Entidades/Restaurante.cs
using AnaliseDeSentimento.Arquivo;
using AnaliseDeSentimento.Constantes;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AnaliseDeSentimento.Entidades
{
    public class Restaurante
    {
        public string Nome { get; init; }
        public decimal Estrelas { get; init; }
        public string Avaliacao { get; init; }

        public EnumClassificacao Classificacao { get; private set; }
        public Dictionary<string, int> Termos { get; private set; }
        public int TermosPositivos { get; private set; }
        public int TermosNegativos { get; private set; }


        public Restaurante(ValueTuple<string, decimal, string> tuple)
        {
            var (nome, estrelas, avaliacao) = tuple;

            Nome = nome;
            Estrelas = estrelas;
            Avaliacao = avaliacao;
            Termos = new();
        }

        public Restaurante()
        {
            Termos = new();
        }

        public static List<Restaurante> ListaDeRestaurantes(LerArquivo arquivo)
            => arquivo.Linhas.Select(linha => new Restaurante(SeparaLinha(linha))).ToList();

        static ValueTuple<string, decimal, string> SeparaLinha(string linha)
        {
            var dadosRestaurante = linha.Split(";");

            var nome = dadosRestaurante[0];
            var estrelas = decimal.Parse(dadosRestaurante[1].Replace('.', ','));
            var avaliacao = dadosRestaurante[2];

            return new(nome, estrelas, avaliacao);
        }

        public void ProcessarTermosAvaliacao(Termos termos)
        {
            var avaliacaoTemp = AvaliacaoTemporaria();
            termos.Items.ForEach((termoAtual) =>
            {
                while (avaliacaoTemp.Contains(termoAtual))
                {
                    AdicionarTermo(termoAtual, termos.Tipo);
                    avaliacaoTemp.Remove(termoAtual);
                }
            });

            Classificacao = ClassificarAvaliacao();
        }

        EnumClassificacao ClassificarAvaliacao()
            => (TermosPositivos - TermosNegativos) switch
            {
                > 0 => EnumClassificacao.Positiva,
                0 => EnumClassificacao.Neutra,
                < 0 => EnumClassificacao.Negativa
            };

        IList<string> AvaliacaoTemporaria()
            => Avaliacao
                .ToLower()
                .Split(' ', '!', '.', ',', ':', ';')
                .ToList();

        void AdicionarTermo(string termo, EnumTermos tipo)
        {
            var termoExiste = Termos.TryGetValue(termo, out int quantidadeEncontrada);
            if (termoExiste)
                Termos.Remove(termo);

            Termos.Add(termo, quantidadeEncontrada += 1);

            AdicionarSomatorioTermo(tipo);
        }

        void AdicionarSomatorioTermo(EnumTermos tipo)
        {
            if (EnumTermos.Positivo == tipo)
                TermosPositivos += 1;

            if (EnumTermos.Negativo == tipo)
                TermosNegativos += 1;
        }

        string ExibirTermos()
            => Termos
                .Select(termo => $"{termo.Key}: {termo.Value}; ")
                .Aggregate(string.Empty, (acumulador, atual) => $"{acumulador}{atual}");

        public override string ToString()
            => $"{Nome}\t{TermosPositivos}\t{TermosNegativos}\t{ExibirTermos()}";
    }
}

File: Python/Entidades/Restaurante.py
import re
import functools
from Arquivo.LerArquivo import LerArquivo
from Entidades.Termos import Termos
from Constantes.EnumTermos import EnumTermos
from Constantes.EnumClassificacao import EnumClassificacao


class Restaurante:
  def __init__(self, valueTuple):
    nome, estrelas, avaliacao = valueTuple

    self.Nome = nome
    self.Estrelas = estrelas
    self.Avaliacao = avaliacao
    self.Termos = dict()
    self.TermosPositivos = 0
    self.TermosNegativos = 0
    self.Classificacao = EnumClassificacao.Neutra

  @staticmethod
  def ListaDeRestaurantes(arquivo: LerArquivo):
    return list(map(lambda linha: Restaurante(Restaurante.SeparaLinha(linha)), arquivo.Linhas))

  @staticmethod
  def SeparaLinha(linha):
    (nome, estrelasString, avaliacao) = linha.split(';')

    estrelas = float(estrelasString)

    return (nome, estrelas, avaliacao)
   

  def AvaliacaoTemporaria(self):
    return re.split(' |,|\.|\!|\\|;|:', self.Avaliacao.lower())
    
  def AdicionarTermo(self, termoAtual, tipoTermo: EnumTermos):
    existeNoDicionario = termoAtual in self.Termos
    quantidadeTermos = 0

    if(existeNoDicionario):
      quantidadeTermos = self.Termos.get(termoAtual)
      self.Termos.pop(termoAtual)
    
    self.Termos[termoAtual] = quantidadeTermos
    self.AdicionarSomatorioTermo(tipoTermo)

  def AdicionarSomatorioTermo(self, tipoTermo: EnumTermos):
    if tipoTermo is EnumTermos.Positivo:
      self.TermosPositivos += 1

    if tipoTermo is EnumTermos.Negativo:
      self.TermosNegativos += 1

  def ProcessarTermosAvaliacao(self, termos: Termos):
    avaliacaoTemp = self.AvaliacaoTemporaria()

    for termoAtual in termos.Items:
      while termoAtual in avaliacaoTemp:
        self.AdicionarTermo(termoAtual, termos.Tipo)
        avaliacaoTemp.remove(termoAtual)

    self.Classificacao = self.ClassificarAvaliacao()

  def ClassificarAvaliacao(self):
    diferencaDeTermosPositivos = self.TermosPositivos - self.TermosNegativos

    if diferencaDeTermosPositivos > 0:
      return EnumClassificacao.Positiva
      
    if diferencaDeTermosPositivos == 0:
      return EnumClassificacao.Neutra

    if diferencaDeTermosPositivos < 0:
      return EnumClassificacao.Negativa

  def __ListarTermos(self):
    resposta = ''
    for key, value in self.Termos.items():
      resposta += "%s: %d; " % (key, value)

    return resposta

  def __str__(self):
    return "%s\t%d\t%d\t%s" % (self.Nome, self.TermosPositivos, self.TermosNegativos, self.__ListarTermos())
